manifests/registry.pp

```diff
@@ -33,19 +33,19 @@
 #   Required to be true for idempotency
 #
 define docker::registry(
-  $server      = $title,
-  $ensure      = 'present',
-  $username    = undef,
-  $password    = undef,
-  $pass_hash   = undef,
-  $email       = undef,
-  $local_user  = 'root',
-  $version     = $docker::version,
-  $receipt     = true,
+  Optional[String] $server      = $title,
+  Optional[String] $ensure      = 'present',
+  Optional[String] $username    = undef,
+  Optional[String] $password    = undef,
+  Optional[String] $pass_hash   = undef,
+  Optional[String] $email       = undef,
+  Optional[String] $local_user  = 'root',
+  Optional[String] $version     = $docker::version,
+  Optional[Boolean] $receipt     = true,
 ) {
   include docker::params
 
-  validate_re($ensure, '^(present|absent)$')
+  assert_type(Pattern[/^present$|^absent$/], $ensure)
```

Comment:
You can use Pattern[] as a data type! So you could remove this line and have the definition further up as -
 Optional[Pattern[/^present$|^absent$/]] $ensure - https://puppet.com/docs/puppet/5.3/lang_data_abstract.html#pattern